pkg/logql/syntax/ast.go

```diff
@@ -37,6 +37,7 @@ func (implicit) logQLExpr() {}
 // LogSelectorExpr is a LogQL expression filtering and returning logs.
 type LogSelectorExpr interface {
 	Matchers() []*labels.Matcher
+	WithMatchers([]*labels.Matcher) LogSelectorExpr
```

Comment:
I wonder if this is some scope creep. The sharding mappers use `Walk` instead. Should we open the door for AST manipulation here?